File: treino/p0028.cpp
#include <algorithm>
#include <bits/stdc++.h>
using namespace std;

typedef vector<int> vi;
typedef map<int, vi> mapivi;
typedef pair<int, int> pii;
typedef vector<pii> vpii;

void getInput(int N,
    int Q,
    vi &planetas,
    vpii &versoes,
    mapivi &origens
  ) {
  for (int i = 0; i < N; i++) {
    int energia;
    cin >> energia;
    planetas[i] = energia;

    if (origens.find(energia) == origens.end()) {
      vi origem(1, i);
      origens[energia] = origem;
    } else {
      origens[energia].push_back(i);
    }
  }

  for (int i = 0; i < Q; i++) {
    pii versao;
    cin >> versao.first >> versao.second;
    versoes[i] = versao;
  }
};

void solve() {
  int N, Q;
  cin >> N >> Q;
  vi planetas(N, 0);
  vpii versoes(Q);
  mapivi origens;

  getInput(N, Q, planetas, versoes, origens);

  for (auto versao: versoes) {
    vi origem_index_list = origens[versao.first];
    vi destino_index_list = origens[versao.second];
    int maior = 0;

    for (auto origem_index : origem_index_list) {
      for (auto destino_index : destino_index_list) {
        int distancia = (destino_index - origem_index +N) % N+1;
        if ( distancia == 0 )
          distancia = N;
        maior = max(maior, distancia);
      }
    }

    cout << maior << "\n";
  }
}

int main() {
  cin.tie(0); cout.tie(0);
  ios::sync_with_stdio(0);
  solve();
  return 0;
}

File: treino/p0028.py
def findIndexes(lista):
    indexes = {}

    for i, v in enumerate(lista):
        if v in indexes:
            indexes[v].append(i)
        else:
            indexes[v] = [i]

    return indexes


N, Q = map(int, input().split())
planetas = list(map(int, input().split()))
versoes = []

for _ in range(Q):
    a, b = map(int, input().split())
    versoes.append((a, b))

origens = findIndexes(planetas)

# for a, b in versoes:
#
#     maior = 0
#
#     for origem in origens[a]:
#         distancia = 1
#         for j in range(origem + 1, origem + N):
#             distancia += 1
#             destino = j % N
#
#             if planetas[destino] == b:
#                 maior = max(maior, distancia)

for a, b in versoes:
    if a not in origens or b not in origens:
        print(0)
        continue

    origem_index_list = origens[a]
    destino_index_list = origens[b]

    maior = 0

    for origem_index in origem_index_list:
        for destino_index in destino_index_list:
            distancia = (destino_index - origem_index + 1) % N
            if distancia == 0:
                distancia = N
            maior = max(maior, distancia)

    print(maior)
